OpenRA.Mods.Common/Traits/World/PathFinder.cs

```diff
@@ -125,95 +115,168 @@ public List<CPos> FindPath(PathSearch search)
 
-					if (path != null)
-						return path;
-				}
+		readonly World world;
+		public NonCachedPathFinder(World world) { this.world = world; }
 
-				// no path exists
-				return EmptyPath;
+		public List<CPos> FindUnitPath(CPos from, CPos target, Actor self)
+		{
+			var mi = self.Info.Traits.Get<MobileInfo>();
+
+			// If a water-land transition is required, bail early
+			var domainIndex = self.World.WorldActor.TraitOrDefault<DomainIndex>();
+			if (domainIndex != null)
+			{
+				var passable = mi.GetMovementClass(world.TileSet);
+				if (!domainIndex.IsPassable(from, target, (uint)passable))
+					return EmptyPath;
 			}
+
+			var pb = FindBidiPath(
+				PathSearch.FromPoint(world, mi, self, target, from, true),
+				PathSearch.FromPoint(world, mi, self, from, target, true).Reverse());
+
+			CheckSanePath2(pb, from, target);
+
+			return pb;
 		}
 
-		static List<CPos> MakePath(CellLayer<CellInfo> cellInfo, CPos destination)
+		public List<CPos> FindUnitPathToRange(CPos src, SubCell srcSub, WPos target, WRange range, Actor self)
 		{
-			var ret = new List<CPos>();
-			var pathNode = destination;
+			var mi = self.Info.Traits.Get<MobileInfo>();
+			var targetCell = self.World.Map.CellContaining(target);
+			var rangeSquared = range.Range * range.Range;
+
+			// Correct for SubCell offset
+			target -= self.World.Map.OffsetOfSubCell(srcSub);
+
+			// Select only the tiles that are within range from the requested SubCell
+			// This assumes that the SubCell does not change during the path traversal
+			var tilesInRange = world.Map.FindTilesInCircle(targetCell, range.Range / 1024 + 1)
+				.Where(t => (world.Map.CenterOfCell(t) - target).LengthSquared <= rangeSquared
+					   && mi.CanEnterCell(self.World, self, t));
+
+			// See if there is any cell within range that does not involve a cross-domain request
+			// Really, we only need to check the circle perimeter, but it's not clear that would be a performance win
+			var domainIndex = self.World.WorldActor.TraitOrDefault<DomainIndex>();
+			if (domainIndex != null)
+			{
+				var passable = mi.GetMovementClass(world.TileSet);
+				tilesInRange = new List<CPos>(tilesInRange.Where(t => domainIndex.IsPassable(src, t, (uint)passable)));
+				if (!tilesInRange.Any())
+					return EmptyPath;
+			}
 
-			while (cellInfo[pathNode].Path != pathNode)
+			var path = FindBidiPath(
+				PathSearch.FromPoints(world, mi, self, tilesInRange, src, true),
+				PathSearch.FromPoint(world, mi, self, src, targetCell, true).Reverse());
+
+			return path;
+		}
+
+		public List<CPos> FindPath(PathSearch search)
+		{
+			using (search)
 			{
-				ret.Add(pathNode);
-				pathNode = cellInfo[pathNode].Path;
+				List<CPos> path = null;
+
+				while (!search.OpenQueue.Empty)
+				{
+					var p = search.Expand(world);
+					if (search.Heuristic(p) == 0)
+					{
+						path = MakePath(search.CellInfo, p);
+						break;
+					}
+				}
+
+				var dbg = world.WorldActor.TraitOrDefault<PathfinderDebugOverlay>();
```

Comment:
You should not take everything that is currently in bleed as good on belief ;)
Unfortunately a lot needs to be rewritten, and since you're refactoring that I figured you might as well fix it.